contracts/ssnlist.scilla

```diff
@@ -1542,3 +1561,538 @@ transition DrainContractBalance(amt: Uint128, initiator: ByStr20)
+          ssn_deleg_amt[ssn_address][new_deleg_addr] := ssn_deleg_deposit_amt;
+          delete ssn_deleg_amt[ssn_address][initial_deleg];
+          e = { _eventname: "SwapSSNDelegAmtNewEntry"; initial_deleg: initial_deleg; new_deleg: new_deleg_addr; ssn_addr: ssn_address; amt: ssn_deleg_deposit_amt };
+          event e
+        end
+      | None => (* no new deleg to transfer to *)
+      end
+    | None => (* No deleg deposit for this ssn *)
+    end
+  | None =>
+    e = DelegDoesNotExistAtSSN;
+    ThrowError e
+  end
+end
+
+(* @dev Swap deleg address from current_deleg -> new_deleg in last_withdraw_cycle_deleg *)
+(*      If the new deleg has an existing withdraw cycle, it is retained. *)
+(*      Otherwise, it is set the the last reward cycle *)
+procedure SwapLastWithdrawCycleDeleg(ssninfo: Pair ByStr20 Ssn)
+  fst = @fst ByStr20 Ssn;
+  ssn_address = fst ssninfo;
+  current_deleg_o <- current_deleg;
+  new_deleg_o <- new_deleg;
+
+  match current_deleg_o with
+  | Some initial_deleg =>
+    lwcd <- last_withdraw_cycle_deleg[initial_deleg][ssn_address];
+    match lwcd with
+    | Some cycle_no =>
+      match new_deleg_o with
+      | Some new_deleg_addr =>
+        new_deleg_lwcd <- last_withdraw_cycle_deleg[new_deleg_addr][ssn_address];
+        match new_deleg_lwcd with
+        | Some existing_lwcd =>
+          (* initial deleg has a last withdraw cycle *)
+          (* recipient has an existing last withdraw cycle *)
+          (* keep the existing lwcd of recipient *)
+          delete last_withdraw_cycle_deleg[initial_deleg][ssn_address]
+        | None =>
+          (* recipient does not have an entry for this ssn *)
+          lrc <- lastrewardcycle;
+          last_withdraw_cycle_deleg[new_deleg_addr][ssn_address] := lrc;
+          delete last_withdraw_cycle_deleg[initial_deleg][ssn_address];
+          e = { _eventname: "SwapLastWithdrawCycleDelegNewEntry"; initial_deleg: initial_deleg; new_deleg: new_deleg_addr; ssn_addr: ssn_address; last_reward_cycle: lrc };
+          event e
+        end
+      | None => (* no new deleg to transfer to *)
+      end
+    | None => (* No last withdraw cycle for this ssn *)
+    end
+  | None =>
+    e = DelegDoesNotExistAtSSN;
+    ThrowError e
+  end
+end
+
+(* @dev Swap deleg address from current_deleg -> new_deleg in last_buf_deposit_cycle_deleg *)
+(*      The last buff deposit cycle of the new deleg is always set to the last reward cycle *)
+procedure SwapLastBuffDepositCycleDeleg(ssninfo: Pair ByStr20 Ssn)
+  fst = @fst ByStr20 Ssn;
+  ssn_address = fst ssninfo;
+  current_deleg_o <- current_deleg;
+  new_deleg_o <- new_deleg;
+
+  match current_deleg_o with
+  | Some initial_deleg =>
+    lbdc <- last_buf_deposit_cycle_deleg[initial_deleg][ssn_address];
+    match lbdc with
+    | Some cycle_no =>
+      match new_deleg_o with
+      | Some new_deleg_addr =>
+        lrc <- lastrewardcycle;
+        last_buf_deposit_cycle_deleg[new_deleg_addr][ssn_address] := lrc;
```

Comment:
why is it lrc instead of lbdc?

Reply:
it is mimicking the `Delegate` procedure. The idea is that we treat it like a delegate. https://github.com/Zilliqa/staking-contract/blob/feat/swap-deleg/contracts/ssnlist.scilla#L961. However, I am not convince if the new assertion (https://github.com/Zilliqa/staking-contract/blob/feat/swap-deleg/contracts/ssnlist.scilla#L2098) that checks for buffered deposit and rewards will let us avoid the issue of not swapping the `lbd` and `lwcd`.